---
title: "fgczgseaora: Unifying methods on gene set enrichment"
author:
- name: Lucas Heinrich Kook
  affiliation: Functional Genomics Center Zurich, ETH Zurich, University of Zurich
- name: Witold Wolski
  affiliation: Functional Genomics Center Zurich, ETH Zurich, University of Zurich
  email: wew@fgcz.ethz.ch
package: fgczgseaora
output:
  BiocStyle::pdf_document
bibliography: bibliography.bib
abstract: |
  A plethora of R packages exist on CRAN and Bioconductor to perform over-representation analysis (ORA) and gene set enrichment analysis (GSEA). However, consistency in the underlying nomenclature for specific analyses and user friendly implementation is still lacking. `fgczgseaora` aims at unifying ID mapping and enrichment analysis in a syntactically coherent and intuitive way, while ensuring reproducibility of results. `fgczgseaora` primarily consists of wrapper functions around the `r CRANpkg("sigora")` and `r CRANpkg("WebGestaltR")` packages from CRAN and `r CRANpkg("rmarkdown")` based reports for visualisation and contextualisation of analysis results.
vignette: |
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
library(tidyverse)
library(fgczgseaora)
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  error = FALSE,
  fig.width = 4,
  fig.height = 3,
  fig.align = "center",
  cache = FALSE
)
options(width = 80)
```


# Introduction

Methods to identify differentially regulated pathways among a vast background of quantified genes or proteins differ in complexity of the underlying algorithms and computations, ranging from the application of Fisher's exact test in ORA and extensions to gene pairs [@Foroushani2013] to more complex statistical procdures in gene set enrichment analysis [@Subramanian15545] and network based procedures [@glaab2012enrichnet]. Almost all methods come with their own implementation as an _R_ package, _i.e._ ORA in both `r CRANpkg("sigora")` and `r CRANpkg("WebGestaltR")`, network based gene enrichment analysis in `r CRANpkg("enrichnet")` and the list keeps going on. Naturally, different implementations come with different -- and often inconsistent -- features, such as ID mapping or demand a lot of knowledge from the user in setting up appropriate data structures and other preconditions. `fgczgseaora` aims at providing a consistent and user friendly interface, which will be introduced in the following section. Elaborating on the methodological and statistical aspects of the underlying methods is beyond the scope of this vignette and the reader is referred to the publications in the reference section.

# Installation and loading

After installing the packages from _Bioconductor_ as shown below, the package can be loaded using `library(fgczgseaora)`.

```{r installing, eval=FALSE}
BiocManager::install("fgczgseaora")
library(fgczgseaora)
```

# Example workflow

We shall analyse proteomic data with the following format:

```{r}
data("exampleContrastData")
glimpse(exampleContrastData)
```

The first column `protein_Id` contains FASTA formatted human protein accessions and the second column fold change estimates. _e.g._ as obtained by a `limma::toptable()` call. The first step is now to translate the FASTA headers into Uniprot accessions for subsequent ID mapping.

```{r}
dd <- getUniprotFromFastaHeader(exampleContrastData)
glimpse(dd)
```

We see that we have lost `r nrow(exampleContrastData) - nrow(dd)` accessions which did not correspond to a FASTA format. Now we apply ORA, sigORA and GSEA to the obtained ranked protein list. To apply sigORA and ORA one must first compute the background GPS repository using `makeGPS_wrappR()`. The advantage is that backgrounds can be specified separately for each experiment or combined for several batches of experiments. However, computing the accession pair signatures and calculating their respective weights is computationally intensive and potentially time consuming.

```{r results='hide'}
data("idmap", package = "sigora")
myGPSrepo <- makeGPS_wrappR(ids = dd$UniprotID, target = "GO")
res <- sigoraWrappR(df = dd, fc_threshold = 0.2, fc_col = "estimate",
                    GPSrepos = myGPSrepo, greater_than = TRUE)
```

The result of `sigoraWrappR()` is a list containing the sigORA and ORA results along with user specified inputs that will be used in the reports described in section \@ref(reports).

```{r}
names(res)
```

Under the hood a lot of ID mapping has taken place of which the user is unaware. To check the ID mapping efficiency when mapping Uniprot IDs to GO, KEGG or ENTREZ accessions the package comes with a `checkIDmappingEfficiency()` function to check the number of lost IDs.

```{r}
checkIDmappingEfficiency(dd$UniprotID, keytype = "UNIPROT") %>% 
  round(2)
```

Evidently, the mapping from Uniprot to KEGG IDs seems to be fairly inefficient and results should be interpreted with care.

# Reports

<!-- <FIXME> Include reports and `.runGSEA()` + other run functions </FIXME> -->

# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References
